File: benchmarks/tree_traversal/r/traversal.R
# Source:  https://github.com/emmanuelparadis/ape (CRAN v5.8-1, 2024-12-16)
# License: GPL-2 | GPL-3 (dual)
#          NOTE: this is a copyleft license; redistribution of code that
#          attaches `ape` inherits GPL terms.
# Adapted: parses a Newick string with `ape::read.tree`, runs a postorder
#          traversal over the `phylo$edge` matrix using the indices returned
#          by `ape::postorder`, and returns aggregate statistics.

#' Traverse a Newick-encoded phylogeny in postorder and return basic stats.
#'
#' @param newick character scalar containing a Newick-encoded tree.
#'
#' @return a named list with components
#'   \describe{
#'     \item{n_tips}{number of leaf tips, from `ape::Ntip`.}
#'     \item{n_internal_nodes}{number of internal nodes, from `ape::Nnode`.}
#'     \item{n_edges}{number of edges visited (== `nrow(tree$edge)`).}
#'     \item{total_branch_length}{sum of `tree$edge.length` over the
#'       postorder traversal.}
#'   }
traverse_newick <- function(newick) {
  stopifnot(is.character(newick), length(newick) == 1L)

  tree <- ape::read.tree(text = newick)

  total_bl   <- 0
  edge_count <- 0L
  for (i in ape::postorder(tree)) {
    total_bl   <- total_bl + tree$edge.length[i]
    edge_count <- edge_count + 1L
  }

  list(
    n_tips              = ape::Ntip(tree),
    n_internal_nodes    = ape::Nnode(tree),
    n_edges             = edge_count,
    total_branch_length = total_bl
  )
}

File: benchmarks/tree_traversal/cpp/traversal.cpp
// Source:  https://github.com/niemasd/CompactTree
//          (main @ 80bdf87d98fb770e37d0298d2aa56d4782fe5081, 2025-03-09)
// License: GPL-3.0
//          NOTE: this is a copyleft license; binaries linking compact_tree.h
//          inherit GPL-3 obligations on redistribution.
// Adapted: parses a Newick string into a `compact_tree`, runs a postorder
//          traversal, and accumulates total node count and summed branch
//          length into a small struct returned to the caller.

#include <cstddef>
#include <string>

#include "compact_tree.h"

namespace thesis_tree
{

struct TreeStats
{
    std::size_t num_nodes;
    double      total_branch_length;
};

/*!\brief Parses a Newick string and returns aggregate traversal statistics.
 *
 * The second constructor argument `is_fn = false` tells CompactTree that the
 * input is a Newick literal, not a filename. Forgetting that flag causes
 * CompactTree to attempt to open the string as a file path.
 */
inline TreeStats traverse_newick(std::string const & newick)
{
    compact_tree tree(newick, /*is_fn=*/false);

    TreeStats stats{0, 0.0};
    for (auto it = tree.postorder_begin(); it != tree.postorder_end(); ++it)
    {
        CT_NODE_T const node = *it;
        stats.num_nodes += 1;
        stats.total_branch_length += static_cast<double>(tree.get_edge_length(node));
    }
    return stats;
}

} // namespace thesis_tree
